File: src/1. Layers/1.3 Domain/Well.Domain/ValueObjects/DeliveryLine.cs
﻿namespace PH.Well.Domain.ValueObjects
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using Enums;

    public class DeliveryLine
    {
        public DeliveryLine()
        {
            Damages = new List<Damage>();
        }

        public int JobDetailId { get; set; }

        public int JobId { get; set; }

        public int LineNo { get; set; }

        public string ProductCode { get; set; }

        public string ProductDescription { get; set; }

        public decimal Value { get; set; }

        public int InvoicedQuantity { get; set; }

        public int ShortQuantity { get; set; }

        public int DeliveredQuantity { get; set; }

        public int ShortsActionId { get; set; }

        public DeliveryAction ShortsAction => (DeliveryAction)ShortsActionId;

        public string LineDeliveryStatus { get; set; }

        public string Reason { get; set; }

        public string Status { get; set; }

        public List<Damage> Damages { get; set; }

        public int JobDetailReasonId { get; set; }

        public int JobDetailSourceId { get; set; }

        public int DamagedQuantity => Damages.Sum(d => d.Quantity);

        public bool IsClean => !IsException;

        private bool IsException => (DamagedQuantity + ShortQuantity > 0); 

        public bool CanSubmit => (ShortQuantity == 0 || ShortsAction != DeliveryAction.NotDefined) &&
                                 Damages.All(d => d.Quantity == 0 || d.DamageAction != DeliveryAction.NotDefined);

        public decimal CreditValueForThreshold()
        {
            var c = (this.ShortQuantity + DamagedQuantity) * this.Value;
            return c;
        }

        public bool IsHighValue { get; set; }

        public bool HasNoActions => ShortsAction == DeliveryAction.NotDefined &&
                                    Damages.All(d => d.DamageAction == DeliveryAction.NotDefined);

    }
}

File: src/1. Layers/1.1 Presentation/Well.Dashboard/app/delivery/deliveryLine.ts
import {Damage} from './damage';

export class DeliveryLine {
    constructor(line: DeliveryLine) {
        if (line) {
            this.jobId = line.jobId;
            this.lineNo = line.lineNo;
            this.productCode = line.productCode;
            this.productDescription = line.productDescription;
            this.value = line.value;
            this.invoicedQuantity = line.invoicedQuantity;
            this.deliveredQuantity = line.deliveredQuantity;
            this.damagedQuantity = line.damagedQuantity;
            this.shortQuantity = line.shortQuantity;

            if (line.damages) {
                var index: number = 0;
                for (let damage of line.damages) {
                    this.damages.push(new Damage(index, damage.quantity, damage.reasonCode));
                    index++;
                }
            }

            this.isCleanOnInit = this.isClean();
        }
    }

    jobId: number;
    lineNo: number;
    productCode: string;
    productDescription: string;
    value: string;
    invoicedQuantity: number;
    deliveredQuantity: number;
    damagedQuantity: number;
    shortQuantity: number;
    damages: Damage[] = new Array<Damage>();

    isCleanOnInit: boolean;

    isClean(): boolean {
        if (this.shortQuantity > 0) {
            return false;
        }

        for (let damage of this.damages) {
            if (damage.quantity > 0) {
                return false;
            }
        }

        return true;
    }
}